### DISCUSSION TITLE
tryEnd not sending FIN packet ?

### DISCUSSION THREAD

**Discussion by Shiranuit** [uNetworking/uWebSockets.js#698] [Q&A] [answered]
Hi, I was inspecting packet using wireshark and found that sending chunks of data using `tryEnd` lead to the FIN packet not being sent,
and since after everything has been sent the response is marked as `done`, it is impossible to force send the FIN packet in any way.
This causes the connection being open until the IDLE Timeout clears it.

Is there a way to close the connection after `tryEnd` ?

Thanks

**Snippet**
```js
const uWS = require('uWebSockets.js');

const port = 7512;

const app = uWS.App({
  
}).get('/tryEnd', (res, req) => {
  // Does not send FIN packet, only IDLE Timeout clears the connection
  res.tryEnd('Hello world');
}).get('/end', (res, req) => {
  res.end('Hello world', true);
}).listen(port, (token) => {
  if (token) {
    console.log('Listening to port ' + port);
  } else {
    console.log('Failed to listen to port ' + port);
  }
});
```

**uWebsocket.js : 20.6.0**

**Comment by e3dio** [comment]
You are missing the 2nd tryEnd parameter totalSize

**Reply by Shiranuit** [reply]
Yes I did it on purpose because when the parameter totalSize is not given `tryEnd` assumes that the whole chunk length is the total length to send.
Adding the second parameter `totalSize` does not change the FIN packet not being sent.

**Reply by uasan** [reply]
But, for HTTP/1.1 this is normal behavior, it should not close connections, it should only end the HTTP request

**Reply by Shiranuit** [reply]
Right, but since the Response is marked as `done` we cannot  force the connection being closed after the `tryEnd`, this actually causes some issue with the Azure Gateway since the connection remains open until the IDLE Timeout.

**Reply by uasan** [reply]
Response as done != Connection as done
HTTP/1.1 assumes that the client closes the connection, not the server.

**Comment by e3dio** [comment]
end() got a closeConnection parameter added recently but tryEnd did not get it, it can be added to tryEnd also

**Comment by uasan** [comment, marked as answer]
It looks like bad design, you want to close connections to Azure Gateway, instead of reusing one connection for many clients, open connections should remain in Azure Gateway free connection pool, that's the point of load balancing

**Reply by Shiranuit** [reply]
I agree, but we were discussing with Azure about that, because their Gateway was sending the error 502 and they were saying that the connection was not properly closed on both sides, that's why I was asking if it was possible to force close the connection on the server side after a try end.
There has been quite a lot of back and forth with them to try understand what could cause this behaviour, and they said that a FIN packet  was not sent from the server side causing the connection to be partially open.

**Reply by uasan** [reply]
> Gateway was sending the error 502

Looks like Azure is trying to work with your backend by HTTP/1.0 protocol.
You can of course degrade your backend to enforce HTTP/1.0 protocol behaviors.

**Reply by ghost** [reply]
There is no FIN in HTTP/1.1 unless client sends Connection: Close, which is a really bad way to send requests - esp. when you have a proxy. The proxy should keep the connection open to minimize overhead
